sgkit/stats/hwe.py

```diff
@@ -0,0 +1,178 @@
+hardy_weinberg_p_value_jit = njit(hardy_weinberg_p_value, fastmath=True)
+
+
+def hardy_weinberg_p_value_vec(
+    obs_hets: ndarray, obs_hom1: ndarray, obs_hom2: ndarray
+) -> ndarray:
+    arrs = [obs_hets, obs_hom1, obs_hom2]
+    if len(set(map(len, arrs))) != 1:
+        raise ValueError("All arrays must have same length")
+    if list(set(map(lambda x: x.ndim, arrs))) != [1]:
+        raise ValueError("All arrays must be 1D")
+    n = len(obs_hets)
+    p = np.empty(n, dtype=np.float64)
+    for i in range(n):
+        p[i] = hardy_weinberg_p_value_jit(obs_hets[i], obs_hom1[i], obs_hom2[i])
+    return p
+
+
+hardy_weinberg_p_value_vec_jit = njit(hardy_weinberg_p_value_vec, fastmath=True)
+
+
+def hardy_weinberg_test(
+    ds: Dataset, genotype_counts: Optional[Hashable] = None
+) -> Dataset:
+    """Exact test for HWE as described in Wigginton et al. 2005 [1].
+
+    Parameters
+    ----------
+    ds : Dataset
+        Dataset containing genotype calls or precomputed genotype counts.
+    genotype_counts : Optional[Hashable], optional
+        Name of variable containing precomputed genotype counts, by default
+        None. If not provided, these counts will be computed automatically
+        from genotype calls. If present, must correspond to an (`N`, 3) array
+        where `N` is equal to the number of variants and the 3 columns contain
+        heterozygous, homozygous reference, and homozygous alternate counts
+        (in that order) across all samples for a variant.
+
+    Warnings
+    --------
+    This function is only applicable to diploid, biallelic datasets.
+
+    Returns
+    -------
+    Dataset
+        Dataset containing (N = num variants):
+        variant_hwe_p_value : (N,) ArrayLike
+            P values from HWE test for each variant as float in [0, 1].
+
+    References
+    ----------
+    - [1] Wigginton, Janis E., David J. Cutler, and Goncalo R. Abecasis. 2005.
+        “A Note on Exact Tests of Hardy-Weinberg Equilibrium.” American Journal of
+        Human Genetics 76 (5): 887–93.
+
+    Raises
+    ------
+    NotImplementedError
+        * If ploidy of provided dataset != 2
+        * If maximum number of alleles in provided dataset != 2
+    """
+    if ds.dims["ploidy"] != 2:
+        raise NotImplementedError("HWE test only implemented for diploid genotypes")
+    if ds.dims["alleles"] != 2:
+        raise NotImplementedError("HWE test only implemented for biallelic genotypes")
+    # Use precomputed genotype counts if provided
+    if genotype_counts is not None:
+        obs = list(da.asarray(ds[genotype_counts]).T)
+    # Otherwise compute genotype counts from calls
+    else:
+        # TODO: Use API genotype counting function instead, e.g.
+        # https://github.com/pystatgen/sgkit/issues/29#issuecomment-656691069
+        mask = ds["call_genotype_mask"].any(dim="ploidy")
+        gtc = xr.where(mask, -1, ds["call_genotype"].sum(dim="ploidy"))  # type: ignore[no-untyped-call]
```

Comment:
Style nit/question: we are being a bit inconsistent on case for these variables. E.g. in `count_alleles` the variables are `G`, `CTS`, etc. Also, in the [summary stats PR](https://github.com/pystatgen/sgkit/pull/102/files) I used `G` for genotypes, but `n_het` for heterogeneous counts. 

Are there some rules we can use? E.g. use lowercase except when we are replicating a paper which uses `X`, etc?

Reply:
https://github.com/pystatgen/sgkit/pull/76/commits/466859e584c2dbe716f92d417c87c150ffa1d278#diff-2ec013d1f086d9558934c38ac90411d1R173

I like the capital convention so array names don't conflict with scalars when there are a lot of both.  When there aren't a lot of either, switching to lower case name sounds good and is what I would prefer too.  I don't know how to make a threshold for that clear, so I've been trying to err on the side of sticking with the capital letter convention for consistency (thanks for flagging this one).

I filed https://github.com/pystatgen/sgkit/issues/117 to make a record of this.